### ISSUE TITLE
Investigate how incorrectly encrypted messages are generated in the cluster

### ISSUE THREAD

**Issue by Selutario** [wazuh/wazuh#15006]
## Description

The logs below were found multiple times in cluster environments where there is a heavy workload (for example, when the master is sending many GB of files to multiple nodes simultaneously).
```
2022/09/16 16:30:57 ERROR: [Worker wazuh-manager-worker-0] [Main] Internal error processing request 'b'syn_m_c_e'': Error 3027 - Unknown received task name: b511ef98-faab-4f52-8e44-5a76f1359ba8
2022/09/16 16:30:57 INFO: [Worker wazuh-manager-worker-0] [Agent-info sync] Starting.
2022/09/16 16:31:02 ERROR: [Worker wazuh-manager-worker-0] [Main] Error in synchronization process: Error 3027 - Unknown received task name: b511ef98-faab-4f52-8e44-5a76f1359ba8
2022/09/16 16:31:10 INFO: [Worker wazuh-manager-worker-0] [Agent-info sync] Finished in 15.581s (6 chunks updated).
2022/09/16 16:32:00 INFO: [Worker wazuh-manager-worker-0] [Keep Alive] Error sending request: timeout expired.
2022/09/16 16:33:20 INFO: [Worker wazuh-manager-worker-0] [Keep Alive] Error sending request: timeout expired.
2022/09/16 16:34:40 INFO: [Worker wazuh-manager-worker-0] [Keep Alive] Error sending request: timeout expired.
2022/09/16 16:36:00 INFO: [Worker wazuh-manager-worker-0] [Keep Alive] Error sending request: timeout expired.
2022/09/16 16:37:20 INFO: [Worker wazuh-manager-worker-0] [Keep Alive] Error sending request: timeout expired.
2022/09/16 16:37:57 ERROR: [Worker wazuh-manager-worker-0] [Main] Connection closed due to an unhandled error: Error 3025 - Could not decrypt message
  File "/var/ossec/framework/python/lib/python3.9/asyncio/selector_events.py", line 870, in _read_ready__data_received
    self._protocol.data_received(data)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.6-py3.9.egg/wazuh/core/cluster/common.py", line 610, in data_received
    for command, counter, payload, flag_divided in self.get_messages():
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.6-py3.9.egg/wazuh/core/cluster/common.py", line 433, in get_messages
    raise exception.WazuhClusterError(3025)

2022/09/16 16:38:07 INFO: [Local Server] [Main] Serving on /var/ossec/queue/cluster/c-internal.sock
2022/09/16 16:38:08 INFO: [Worker wazuh-manager-worker-0] [Main] The master closed the connection
2022/09/16 16:38:18 INFO: [Local Server] [Main] Serving on /var/ossec/queue/cluster/c-internal.sock
2022/09/16 16:38:19 INFO: [Worker wazuh-manager-worker-0] [Main] The master closed the connection
2022/09/16 16:38:29 INFO: [Local Server] [Main] Serving on /var/ossec/queue/cluster/c-internal.sock
2022/09/16 16:38:29 INFO: [Worker wazuh-manager-worker-0] [Main] The master closed the connection
2022/09/16 16:38:39 INFO: [Local Server] [Main] Serving on /var/ossec/queue/cluster/c-internal.sock
2022/09/16 16:38:40 INFO: [Worker wazuh-manager-worker-0] [Main] The master closed the connection
2022/09/16 16:38:50 INFO: [Local Server] [Main] Serving on /var/ossec/queue/cluster/c-internal.sock
2022/09/16 16:38:51 INFO: [Worker wazuh-manager-worker-0] [Main] The master closed the connection
2022/09/16 16:39:01 INFO: [Local Server] [Main] Serving on /var/ossec/queue/cluster/c-internal.sock
```

The multiple reconnections that can be seen after `2022/09/16 16:38:07` are related to https://github.com/wazuh/wazuh/issues/14996. However, the question is why an incorrectly encrypted message arrived. In the logs for this environment, the issue occurred twice, both times affecting the same worker: `Worker wazuh-manager-worker-0`.

## Checks

The following elements have been updated or reviewed (should also be checked if no modification is required):
- [ ] Tests (unit tests, API integration tests).
- [ ] Changelog.
- [ ] Documentation.
- [ ] Integration test mapping (using `api/test/integration/mapping/_test_mapping.py`).

**Comment by Selutario** [comment]
## Repeated `Error sending request` messages

If a cluster node (in this case a worker) receives a message that does not follow the structure established within the [msg_build()](https://github.com/wazuh/wazuh/blob/e2b514bef3d148acd4bcae1a1c7fa8783b82ca3a/framework/wazuh/core/cluster/common.py#L318-L379) function, it will return incorrect values when unpacked using struct:
https://github.com/wazuh/wazuh/blob/e2b514bef3d148acd4bcae1a1c7fa8783b82ca3a/framework/wazuh/core/cluster/common.py#L95

The problem occurs when the length declared in the header is greater than the actual length of the payload. In those cases, the receiving node will wait until receiving all the bytes that it expects:
https://github.com/wazuh/wazuh/blob/e2b514bef3d148acd4bcae1a1c7fa8783b82ca3a/framework/wazuh/core/cluster/common.py#L425-L426

This means that the receiving node can continue sending data to its peer, but it does not mark any of the received data as complete, causing timeout messages for each of the expected messages:
```
2022/09/16 16:32:00 INFO: [Worker wazuh-manager-worker-0] [Keep Alive] Error sending request: timeout expired.
2022/09/16 16:33:20 INFO: [Worker wazuh-manager-worker-0] [Keep Alive] Error sending request: timeout expired.
2022/09/16 16:34:40 INFO: [Worker wazuh-manager-worker-0] [Keep Alive] Error sending request: timeout expired.
2022/09/16 16:36:00 INFO: [Worker wazuh-manager-worker-0] [Keep Alive] Error sending request: timeout expired.
2022/09/16 16:37:20 INFO: [Worker wazuh-manager-worker-0] [Keep Alive] Error sending request: timeout expired.
```

The proof that it is sending messages but not interpreting the response correctly is that its peer, the master node, is not raising any error or warning for not receiving keepalives from worker0. That is, the master receives keepalives from worker and sends a response, but the worker is accumulating bytes from multiple responses before marking it as a complete response (since expected size is larger than the real payload size).

This would also explain why, once the response is "complete", trying to decrypt it fails as it is actually made up of multiple individual packets/responses:
```
2022/09/16 16:37:57 ERROR: [Worker wazuh-manager-worker-0] [Main] Connection closed due to an unhandled error: Error 3025 - Could not decrypt message
  File "/var/ossec/framework/python/lib/python3.9/asyncio/selector_events.py", line 870, in _read_ready__data_received
    self._protocol.data_received(data)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.6-py3.9.egg/wazuh/core/cluster/common.py", line 610, in data_received
    for command, counter, payload, flag_divided in self.get_messages():
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.6-py3.9.egg/wazuh/core/cluster/common.py", line 433, in get_messages
    raise exception.WazuhClusterError(3025)
```

**Comment by Selutario** [comment]
## Status update

I have continued investigating logs and testing on a 4.3.7 cluster, although so far I have not been able to find the source of the encryption or packet-creation problem.

It is possible to find the same error in multiple days for the same environment, and in all of them, before raising the `Could not decrypt message` exception, the worker get timeouts for every expected response, although it still able to send commands and data which is also still processed by the master:
```
2022/09/12 14:21:18 INFO: [Worker wazuh-manager-worker-2] [Agent-info sync] Finished in 13.739s (2 chunks updated).
2022/09/12 14:21:28 INFO: [Worker wazuh-manager-worker-2] [Agent-info sync] Starting.
2022/09/12 14:21:30 INFO: [Worker wazuh-manager-worker-2] [Keep Alive] Sucessful response from master: keepalive
2022/09/12 14:21:35 INFO: [Worker wazuh-manager-worker-2] [Agent-info sync] Finished in 8.242s (2 chunks updated).
2022/09/12 14:21:47 INFO: [Worker wazuh-manager-worker-2] [Agent-info sync] Starting.
2022/09/12 14:22:34 ERROR: [Worker wazuh-manager-worker-2] [Main] Error 3039 - Timeout while waiting to receive a file: Integrity sync at wazuh-manager-worker-2
2022/09/12 14:22:50 INFO: [Worker wazuh-manager-worker-2] [Keep Alive] Error sending request: timeout expired.
2022/09/12 14:24:10 INFO: [Worker wazuh-manager-worker-2] [Keep Alive] Error sending request: timeout expired.
2022/09/12 14:25:30 INFO: [Worker wazuh-manager-worker-2] [Keep Alive] Error sending request: timeout expired.
2022/09/12 14:26:50 INFO: [Worker wazuh-manager-worker-2] [Keep Alive] Error sending request: timeout expired.
2022/09/12 14:28:10 INFO: [Worker wazuh-manager-worker-2] [Keep Alive] Error sending request: timeout expired.
2022/09/12 14:29:30 INFO: [Worker wazuh-manager-worker-2] [Keep Alive] Error sending request: timeout expired.
2022/09/12 14:30:50 INFO: [Worker wazuh-manager-worker-2] [Keep Alive] Error sending request: timeout expired.
2022/09/12 14:32:10 INFO: [Worker wazuh-manager-worker-2] [Keep Alive] Error sending request: timeout expired.
2022/09/12 14:32:50 ERROR: [Worker wazuh-manager-worker-2] [Main] Connection closed due to an unhandled error: Error 3025 - Could not decrypt message
  File "/var/ossec/framework/python/lib/python3.9/asyncio/selector_events.py", line 870, in _read_ready__data_received
    self._protocol.data_received(data)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.6-py3.9.egg/wazuh/core/cluster/common.py", line 610, in data_received
    for command, counter, payload, flag_divided in self.get_messages():
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.6-py3.9.egg/wazuh/core/cluster/common.py", line 433, in get_messages
    raise exception.WazuhClusterError(3025)

2022/09/12 14:33:00 INFO: [Local Server] [Main] Serving on /var/ossec/queue/cluster/c-internal.sock
2022/09/12 14:33:02 INFO: [Worker wazuh-manager-worker-2] [Main] The master closed the connection
2022/09/12 14:33:12 INFO: [Local Server] [Main] Serving on /var/ossec/queue/cluster/c-internal.sock
2022/09/12 14:33:13 INFO: [Worker wazuh-manager-worker-2] [Main] The master closed the connection
2022/09/12 14:33:23 INFO: [Local Server] [Main] Serving on /var/ossec/queue/cluster/c-internal.sock
2022/09/12 14:33:24 INFO: [Worker wazuh-manager-worker-2] [Main] The master closed the connection
2022/09/12 14:33:34 INFO: [Local Server] [Main] Serving on /var/ossec/queue/cluster/c-internal.sock
```

I suspect that the problem might have to do with the refactor to `send_request` that was made in the following issue and that allows the cluster to send messages larger than 5MB, which is the default limit. These changes were merged into version v4.3, just when this problem has appeared:
- https://github.com/wazuh/wazuh/issues/6820

However, in my local env, I have not been able to reproduce the bug despite forcing `send_request` to send up to 100MB using the `divide` flag.

**Comment by Selutario** [comment]
## Status update

While trying to reproduce this bug, I ran into some other strange behavior that we had seen sometimes within the loaded environment of this user: an error message with no content.
```
2022/10/04 09:42:12 INFO: [Master] [Local integrity] Finished in 0.011s. Calculated metadata of 171 files.
2022/10/04 09:42:12 INFO: [Worker worker2] [Integrity check] Starting.
2022/10/04 09:42:12 INFO: [Worker worker1] [Integrity check] Starting.
2022/10/04 09:42:12 ERROR: [Worker worker2] [Main]  
2022/10/04 09:42:12 ERROR: [Worker worker1] [Main]  
2022/10/04 09:42:12 ERROR: [Worker worker1] [Main] Internal error processing request 'b'syn_i_w_m_e'': Error 3027 - Unknown received task name: 98be578e-a75a-4228-9ae4-7b82dcfa131f
```
The reason is simply that a `timeoutError` could be raised here, which is not accompanied by any description:
https://github.com/wazuh/wazuh/blob/e2b514bef3d148acd4bcae1a1c7fa8783b82ca3a/framework/wazuh/core/cluster/master.py#L654-L656
So nothing appeared when the message was logged:
https://github.com/wazuh/wazuh/blob/e2b514bef3d148acd4bcae1a1c7fa8783b82ca3a/framework/wazuh/core/cluster/common.py#L236-L239

This problem is fixed in 4.4.0 after the changes introduced here, so there is no need to open new issues.
- https://github.com/wazuh/wazuh/issues/11945

Regarding the error decrypting a message, I have not been able to reproduce the problem in my local environment. Then I have created an environment with 7 workers, similar to the one we tested [here](https://github.com/wazuh/wazuh/issues/14996#issuecomment-1265594561). The difference is that this time I have modified the intervals in the `cluster.json` so that the integrity syncs are much more frequent (every 0.1 seconds). I have also modified other values so that the timeout when receiving files occurs after waiting 1 second. Lastly, I created smaller files (200MB total) so the master node wouldn't run out of disk space during Integrity sync. 

Although the amount of zips that the master was sending simultaneously was increasing (up to 7 zip files per node), no error occurred while decrypting messages in the workers.

## Conclusion

This is an error that only seems to occur in environments with extreme conditions. We have not been able to reproduce it or find the cause in the code, so we cannot continue the investigation.
